libraries/chain/snapshot_scheduler.cpp

```diff
@@ -0,0 +1,254 @@
+   _snapshot_requests.erase(existing);
+   x_serialize();
+
+   // returning snapshot_schedule_result
+   return result;
+}
+
+snapshot_scheduler::get_snapshot_requests_result snapshot_scheduler::get_snapshot_requests() {
+   snapshot_scheduler::get_snapshot_requests_result result;
+   auto& asvector = _snapshot_requests.get<as_vector>();
+   result.snapshot_requests.reserve(asvector.size());
+   result.snapshot_requests.insert(result.snapshot_requests.begin(), asvector.begin(), asvector.end());
+   return result;
+}
+
+void snapshot_scheduler::set_db_path(bfs::path db_path) {
+   _snapshot_db.set_path(std::move(db_path));
+   // init from db
+   if(std::filesystem::exists(_snapshot_db.get_json_path())) {
+      std::vector<snapshot_scheduler::snapshot_schedule_information> sr;
+      _snapshot_db >> sr;
+      // if db read succeeded, clear/load
+      _snapshot_requests.get<by_snapshot_id>().clear();
+      _snapshot_requests.insert(sr.begin(), sr.end());
+   }
+}
+
+void snapshot_scheduler::set_snapshots_path(bfs::path sn_path) {
+   _snapshots_dir = std::move(sn_path);
+}
+
+void snapshot_scheduler::add_pending_snapshot_info(const snapshot_scheduler::snapshot_information& si) {
+   auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+   auto snapshot_req = snapshot_by_id.find(_inflight_sid);
+   if(snapshot_req != snapshot_by_id.end()) {
+      _snapshot_requests.modify(snapshot_req, [&si](auto& p) {
+         p.pending_snapshots.emplace_back(si);
+      });
+   }
+}
+
+void snapshot_scheduler::execute_snapshot(uint32_t srid, chain::controller& chain) {
+   _inflight_sid = srid;
+   auto next = [srid, this](const chain::next_function_variant<snapshot_scheduler::snapshot_information>& result) {
+      if(std::holds_alternative<fc::exception_ptr>(result)) {
+         try {
+            std::get<fc::exception_ptr>(result)->dynamic_rethrow_exception();
+         } catch(const fc::exception& e) {
+            EOS_THROW(snapshot_execution_exception,
+                     "Snapshot creation error: ${details}",
+                     ("details", e.to_detail_string()));
+         } catch(const std::exception& e) {
+            EOS_THROW(snapshot_execution_exception,
+                     "Snapshot creation error: ${details}",
+                     ("details", e.what()));
+         }
+      } else {
+         // success, snapshot finalized
+         auto snapshot_info = std::get<snapshot_scheduler::snapshot_information>(result);
+         auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+         auto snapshot_req = snapshot_by_id.find(srid);
+
+         if(snapshot_req != snapshot_by_id.end()) {
+            _snapshot_requests.modify(snapshot_req, [&](auto& p) {
+               auto& pending = p.pending_snapshots;
+               auto it = std::remove_if(pending.begin(), pending.end(), [&snapshot_info](const snapshot_scheduler::snapshot_information& s) { return s.head_block_num <= snapshot_info.head_block_num; });
```

Comment:
```suggestion
               auto it = std::remove_if(pending.begin(), pending.end(), [&snapshot_info](const snapshot_information& s) { return s.head_block_num <= snapshot_info.head_block_num; });
```

and  more to remove!